test/grouping.jl

```diff
@@ -103,12 +106,10 @@ gd = groupby(df, :g)
 newvar = :n
 
 @testset "Limits of @based_on" begin
-    @test_throws LoadError @eval @based_on(gd, :i)
-    @test @based_on(gd, [:i, :g]) == DataFrame(g = df.g, x1 = df.i, x2 = df.g)
-    @test_throws ArgumentError @eval @based_on(gd, All())
-    @test_throws MethodError @eval @based_on(gd, Between(:i, :t)).Between == df.i
-    @test_throws ArgumentError @eval @based_on(gd, Not(:i)).Not == df.i
-    @test_throws ArgumentError @eval @based_on(gd, Not([:i, :g]))
+    @test @based_on(gd, [:i, :g]).i_g_function isa Vector{<:SubArray}
```

Comment:
Still applies.

Reply:
This is in "limits" because its not behavior anyone expects and I don't want to consider this behavior part of the public API. (Hence "limits" instead of "errors"). Is that feasible?

Will still add a real test for equality and type

Reply:
OK. Though we should quickly turn this either into an error or into supported behavior.

Reply:
Yes. Tentatively, I propose the following rule

1. All expressions of the form `y = f(:x)` get the macro applied to them
2. All expressions that *don't* have that structure get evaluated as if they were in the basic DataFrames pipes, this is so `@select(df, Between(:a, :b))` works. 
3. If users want to have an expression that is *not* of the form `y = f(:x)` but still want to refer to `:x` as the column `df.x`, which only applies in `combine`. Then they have to do 

```
@based_on(df, @fun f(:x))
```

So if that is implemented 

```
@based_on(gd, [:i, :g])
```

would be the exact same as `combine(df, [:i, :g])` and if they really wanted a vector of the `SubArray`s they would have to do

```
@based_on(gd, @fun [:i, :g])
```

This is a very tentative proposal but would realistically get around this problem.

Reply:
I'm not a big fan of special-casing `y = f(:x)`. What if we required people to write `All([:i, :g])` (or `Cols([i, :g])` if we add this) when they want to get these columns? Anyway you can always write `based_on(df, :i, :g)`, which is actually shorter.

(BTW, `@based_on` should really be deprecated in favor of `@combine` for consistency with DataFrames. I've never liked that name anyway.)

Reply:
I disagree. In DataFrames

```
combine(df, [:a, :b], :i, [:x, :y] => fun => :z 
```

The "selectors" mean different things than the `[:x, :y]` in the `Pair`. The `[:a, :b]` mean "keep these columns", whereas the `[:x, :y]` area things that get passed to a function. 

I think special casing `y = f(:x)` is consistent with the way `:x`, or inputs to `Pair`s more generally, is used in base data frames. 

More importantly, though, people want `@select(df, :a, y = f(:x))` to work and special casing `y = f(:x)` makes that feasible. 

Obviously we could `@select(df, :a)` passed into the whole split apply combine infrastructure, but then we would end up with `:a_fun` and have to worry about that.

Reply:
OK. But then it's not really `y = f(x)` that would be special-cased, but just the fact that a keyword argument is passed (equivalent to a pair in DataFrames).

One issue though is that in DataFrames the pair syntax allows quite naturally to omit the output names in the case where you return a named tuple. Requiring `@fun` in DataFramesMeta wouldn't be as nice. That's not high priority, but it would be interesting to find another solution.